constraint/marshal.go

```diff
@@ -0,0 +1,363 @@
+package constraint
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"reflect"
+
+	"github.com/consensys/gnark/internal/backend/ioutils"
+	"github.com/fxamacker/cbor/v2"
+	"golang.org/x/sync/errgroup"
+)
+
+// ToBytes serializes the constraint system to a byte slice
+// This is not meant to be called directly since the constraint.System is embedded in
+// a "curve-typed" system (e.g. bls12-381.system)
+func (cs *System) ToBytes() ([]byte, error) {
+	// we prepare and write 4 distinct blocks of data;
+	// that allow for a more efficient serialization/deserialization (+ parallelism)
+	var calldata, instructions, levels []byte
+	var g errgroup.Group
+	g.Go(func() error {
+		var err error
+		calldata, err = cs.calldataToBytes()
+		return err
+	})
+	g.Go(func() error {
+		var err error
+		instructions, err = cs.instructionsToBytes()
+		return err
+	})
+	g.Go(func() error {
+		var err error
+		levels, err = cs.levelsToBytes()
+		return err
+	})
+	body, err := cs.toBytes()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+
+	// header
+	h := header{
```

Comment:
I see why using `[]byte` now instead of `io.Reader`, this allows to create correct header right now. Maybe instead we can still use `io.Writer` in calls to `..ToBytes` above by writing header first with padded zeros, then seeking back to start and overwriting with actual lengths?

Reply:
But I think we do not need the actual lengths at all? If we use some end-of-struct separator in the serialization, then during Reading in method `System.FromBytes` the methods `levelsFromBytes` etc. can just read until the separator?

Reply:
Or we can prepend the number of elements (not the number of bytes). And in `...FromBytes` we read as many elements expected.

Reply:
having the lengths enable the reader to process in parallel the sections. But yes we could avoid it in few ways, it's just more convenient and is quite cheap (4 uint64 vs the weight of the circuit is nothing).